plugins/weida-raw/fu-weida-raw-device.c

```diff
@@ -0,0 +1,872 @@
+					     firmware,
+					     fu_progress_get_child(progress),
+					     error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_prepare(FuDevice *device,
+			    FuProgress *progress,
+			    FwupdInstallFlags flags,
+			    GError **error)
+{
+	FuWeidaRawDevice *self = FU_WEIDA_RAW_DEVICE(device);
+	/* TODO: anything the device has to do before the update starts */
+	g_assert(self != NULL);
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_cleanup(FuDevice *device,
+			    FuProgress *progress,
+			    FwupdInstallFlags flags,
+			    GError **error)
+{
+	FuWeidaRawDevice *self = FU_WEIDA_RAW_DEVICE(device);
+	if (!fu_weida_raw_w8760_protect_flash(self, FU_WEIDA_RAW_CMD8760U16_PROTECT_ALL, error)) {
+		g_prefix_error(error, "protect all failed: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_write_firmware(FuDevice *device,
+				   FuFirmware *firmware,
+				   FuProgress *progress,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	FuWeidaRawDevice *self = FU_WEIDA_RAW_DEVICE(device);
+
+	/* sanity check: add other weida devices as required */
+	if (self->dev_type != FU_WEIDA_RAW_DEV_TYPE_FW8760) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "device type 0x%x not supported",
+			    (guint)self->dev_type);
+		return FALSE;
+	}
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 35, NULL);
+
+	/* get default image */
+	if (!fu_weida_raw_w8760_write_wif1(self, firmware, fu_progress_get_child(progress), error))
+		return FALSE;
+
+	// FIXME: can this move to detach() or cleanup()?
+	if (!fu_weida_raw_w8760_dev_reboot(self, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	fu_device_sleep(device, 2 * 1000);
+
+	if (!fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_INTERNAL)) {
+		fu_device_set_remove_delay(FU_DEVICE(self), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+		fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	} else {
+		fu_wdt_hid_device_rebind_driver(self, error);
+	}
+
+	/* TODO: verify each block */
+	fu_progress_step_done(progress);
+
+	/* success! */
+	return TRUE;
+}
+
+static gboolean
+fu_weida_raw_device_set_quirk_kv(FuDevice *device,
+				 const gchar *key,
+				 const gchar *value,
+				 GError **error)
+{
+	/* TODO: parse value from quirk file */
+	return TRUE;
+	/* failed */
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "quirk key not supported");
+	return FALSE;
+}
+
+static void
+fu_weida_raw_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
```

Comment:
What numbers does the daemon give you? Once you've got the numbers (red text!) you can remove `FU_PROGRESS_FLAG_GUESSED` to remove the daemon warning.